File: backend/app/storage/models/provider.py
"""
ModelProvider model — LLM provider credentials and configuration.

Stores encrypted API keys for OpenAI, Azure OpenAI, Groq, etc.
Each provider belongs to a single user (owner).
"""

import uuid

from sqlalchemy import Boolean, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.storage.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class ModelProvider(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "model_providers"

    # --- Foreign Keys ---
    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # --- Columns ---
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    provider_type: Mapped[str] = mapped_column(
        String(50), nullable=False
    )  # 'openai', 'azure_openai', 'groq', 'openai_compatible', 'custom'
    encrypted_api_key: Mapped[str] = mapped_column(Text, nullable=False)
    endpoint_url: Mapped[str | None] = mapped_column(
        String(500), nullable=True
    )  # Required for Azure
    model: Mapped[str | None] = mapped_column(
        String(100), nullable=True
    )  # e.g. gpt-4o, llama-3.3-70b-versatile
    is_valid: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    # --- Rate limiting ---
    requests_per_minute: Mapped[int | None] = mapped_column(nullable=True)
    tokens_per_minute: Mapped[int | None] = mapped_column(nullable=True)
    max_calls_per_experiment: Mapped[int | None] = mapped_column(nullable=True)

    # --- Judge / light-task config (separate optional provider) ---
    judge_api_key_enc: Mapped[str | None] = mapped_column(Text, nullable=True)
    judge_endpoint: Mapped[str | None] = mapped_column(Text, nullable=True)
    judge_model: Mapped[str | None] = mapped_column(Text, nullable=True)
    use_separate_judge: Mapped[bool] = mapped_column(Boolean, default=False)

    # --- Default target endpoint and system prompt ---
    default_target_endpoint: Mapped[str | None] = mapped_column(Text, nullable=True)
    default_system_prompt: Mapped[str | None] = mapped_column(Text, nullable=True)

    # --- Backup key failover ---
    backup_api_key_enc: Mapped[str | None] = mapped_column(Text, nullable=True)
    backup_endpoint: Mapped[str | None] = mapped_column(Text, nullable=True)
    backup_model: Mapped[str | None] = mapped_column(Text, nullable=True)
    use_backup_key: Mapped[bool] = mapped_column(Boolean, default=False)

    # --- Relationships ---
    owner: Mapped["User"] = relationship(  # noqa: F821
        "User",
        back_populates="providers",
    )
    experiments: Mapped[list["Experiment"]] = relationship(  # noqa: F821
        "Experiment",
        back_populates="provider",
        lazy="noload",
    )

    def __repr__(self) -> str:
        return f"<ModelProvider id={self.id} name={self.name!r} type={self.provider_type!r}>"

File: frontend/src/types/provider.ts
// ---------------------------------------------------------------------------
// Provider types (matches backend schemas/providers.py)
// ---------------------------------------------------------------------------

import type { UserBrief, PaginatedResponse } from './api';

export type ProviderType =
    | 'openai'
    | 'azure_openai'
    | 'groq'
    | 'anthropic'
    | 'gemini'
    | 'ollama'
    | 'mistral'
    | 'together'
    | 'deepseek'
    | 'perplexity'
    | 'fireworks'
    | 'xai'
    | 'cohere'
    | 'huggingface'
    | 'openai_compatible';

export const PROVIDER_OPTIONS: { value: ProviderType; label: string; needsEndpoint?: boolean; noKey?: boolean }[] = [
    { value: 'openai',           label: 'OpenAI' },
    { value: 'anthropic',        label: 'Anthropic (Claude)' },
    { value: 'gemini',           label: 'Google Gemini' },
    { value: 'groq',             label: 'Groq' },
    { value: 'mistral',          label: 'Mistral AI' },
    { value: 'together',         label: 'Together AI' },
    { value: 'deepseek',         label: 'DeepSeek' },
    { value: 'perplexity',       label: 'Perplexity AI' },
    { value: 'fireworks',        label: 'Fireworks AI' },
    { value: 'xai',              label: 'xAI (Grok)' },
    { value: 'cohere',           label: 'Cohere' },
    { value: 'huggingface',      label: 'HuggingFace' },
    { value: 'azure_openai',     label: 'Azure OpenAI', needsEndpoint: true },
    { value: 'ollama',           label: 'Ollama (Local)', noKey: true },
    { value: 'openai_compatible', label: 'OpenAI-Compatible', needsEndpoint: true },
];

export interface ProviderCreate {
    name: string;
    provider_type: ProviderType;
    api_key: string;
    endpoint_url?: string;
    model?: string;
    // Rate limiting
    requests_per_minute?: number | null;
    tokens_per_minute?: number | null;
    max_calls_per_experiment?: number | null;
    // Judge config
    judge_api_key?: string | null;
    judge_endpoint?: string | null;
    judge_model?: string | null;
    use_separate_judge?: boolean;
    // Default config
    default_target_endpoint?: string | null;
    default_system_prompt?: string | null;
    // Backup key failover
    backup_api_key?: string | null;
    backup_endpoint?: string | null;
    backup_model?: string | null;
    use_backup_key?: boolean;
}

export interface ProviderUpdate {
    name?: string;
    api_key?: string;
    endpoint_url?: string;
    model?: string;
    // Rate limiting
    requests_per_minute?: number | null;
    tokens_per_minute?: number | null;
    max_calls_per_experiment?: number | null;
    // Judge config
    judge_api_key?: string | null;
    judge_endpoint?: string | null;
    judge_model?: string | null;
    use_separate_judge?: boolean;
    // Default config
    default_target_endpoint?: string | null;
    default_system_prompt?: string | null;
    // Backup key failover
    backup_api_key?: string | null;
    backup_endpoint?: string | null;
    backup_model?: string | null;
    use_backup_key?: boolean;
}

export interface Provider {
    id: string;
    name: string;
    provider_type: ProviderType;
    endpoint_url: string | null;
    model: string | null;
    api_key_preview: string;
    is_valid: boolean | null;
    has_experiments: boolean;
    created_by: UserBrief | null;
    created_at: string;
    updated_at: string;
    // Rate limiting
    requests_per_minute?: number | null;
    tokens_per_minute?: number | null;
    max_calls_per_experiment?: number | null;
    // Judge config
    judge_api_key?: string | null;
    judge_endpoint?: string | null;
    judge_model?: string | null;
    use_separate_judge?: boolean;
    // Default config
    default_target_endpoint?: string | null;
    default_system_prompt?: string | null;
    // Backup key failover
    backup_api_key?: string | null;
    backup_endpoint?: string | null;
    backup_model?: string | null;
    use_backup_key?: boolean;
}

export type ProviderList = PaginatedResponse<Provider>;

export interface ProviderValidationResult {
    is_valid: boolean;
    message: string;
    latency_ms: number | null;
}
